rusk/src/lib/chain/rusk.rs

```diff
@@ -509,6 +522,16 @@ fn execute(
         receipt.gas_spent = receipt.gas_limit;
     }
 
+    if let Some(deploy) = tx.payload().contract_deploy() {
+        session.deploy_raw(
+            deploy.contract_id.map(|id| id.into()),
+            deploy.bytecode.bytes.as_slice(),
+            deploy.constructor_args.clone(),
+            deploy.owner.clone(),
+            tx.payload().fee.gas_limit,
+        )?;
```

Comment:
If this fails, the transaction that passed through "spend_and_execute" will be in the state regardless. Are we sure they get rolled back? Either through re-execution or another mechanism?

Reply:
My understanding is (and test confirms that) that transaction gets discarded, and this new state is never accepted. Caveat is that user does not pay anything for such try, yet, should she pay if the tx is discarded?

Reply:
This is true for AST and FST for sure, so we won't be affecting the running network. However, in EST there's logic that assumes the discarded transaction is atomic, which is not the case here anymore.

Reply:
Do you have maybe some solution in mind?

Reply:
To address the above failing deployment scenario, I moved deployment into the loop and reused your re-execution mechanism - please have a look.